src/components/modals/cloud-tile.js

```diff
@@ -0,0 +1,182 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+import React from 'react';
+import styled from 'styled-components';
+import {Logout, Login} from 'components/common/icons';
+import {
+  CenterVerticalFlexbox,
+  Button
+} from 'components/common/styled-components';
+
+const StyledTileWrapper = styled.div`
+  display: flex;
+  flex-direction: column;
+  align-items: center;
+  justify-content: flex-start;
+  border-radius: 2px;
+  border: 1px solid
+    ${props =>
+      props.selected
+        ? props.theme.primaryBtnBgd
+        : props.theme.selectBorderColorLT};
+  color: ${props =>
+    props.selected
+      ? props.theme.primaryBtnBgd
+      : props.theme.selectBorderColorLT};
+  cursor: pointer;
+  font-weight: 500;
+  width: 120px;
+  height: 168px;
+  background-color: #ffffff;
+  transition: ${props => props.theme.transition};
+  position: relative;
+  :hover {
+    border: 1px solid ${props => props.theme.primaryBtnBgd};
+    color: ${props => props.theme.primaryBtnBgd};
+  }
+
+  .button {
+    margin-top: 20px;
+  }
+`;
+
+const CheckMark = styled.span.attrs({
+  className: 'checkbox-inner'
+})`
+  background-color: ${props => props.theme.primaryBtnBgd};
+  position: absolute;
+  bottom: 0;
+  right: 0;
+  display: block;
+  width: 16px;
+  height: 16px;
+  border-top-left-radius: 2px;
+
+  :after {
+    position: absolute;
+    display: table;
+    border: 2px solid #fff;
+    border-top: 0;
+    border-left: 0;
+    transform: rotate(45deg) scale(1) translate(-50%, -50%);
+    opacity: 1;
+    content: ' ';
+    top: 50%;
+    left: 25%;
+    width: 5.7px;
+    height: 9.1px;
+  }
+`;
+const StyledBox = styled(CenterVerticalFlexbox)`
+  margin-right: 12px;
+`;
+
+const StyledCloudName = styled.div`
+  font-size: 12px;
+  margin-top: 12px;
+  margin-bottom: 4px;
+`;
+
+const StyledUserName = styled.div`
+  font-size: 11px;
+  margin-top: 8px;
+  text-align: center;
+  color: ${props => props.theme.primaryBtnActBgd};
+  overflow: hidden;
+  width: 100px;
+  text-overflow: ellipsis;
+`;
+
+const LoginButton = ({onClick}) => (
+  <Button link small onClick={onClick}>
+    <Login />
+    Login
+  </Button>
+);
+
+const LogoutButton = ({onClick}) => (
+  <Button link small onClick={onClick}>
+    <Logout />
+    Logout
+  </Button>
+);
+
+// eslint-disable-next-line complexity
+const CloudTile = ({
+  // action when click on the tile
+  onSelect,
+  // default to login
+  onConnect,
+  // default to logout
+  onLogout,
+  // action name
+  actionName,
+  // cloud provider class
+  cloudProvider,
+  // function to take after login or logout
+  onSetCloudProvider,
+  // whether provider is selected as currentProvider
+  isSelected,
+  // whether user has logged in
+  isConnected
+}) => {
+  const userName =
```

Comment:
we can improve performance by avoiding recomputing for every rendering

```
import {useMemo} from 'react';
const userName = useMemo(() =>  
  typeof cloudProvider.getUserName === 'function'
      ? cloudProvider.getUserName()
      : null
, [cloudProvider.getUserName]);
```

Reply:
we need to specifically recomputing because cloudProvider is a class stored in the store, changing it will not cause react re render

Reply:
because we are calling getUserName every time this component renders, we could memoize its value by using useMemo

Reply:
`useMemo will only recompute the memoized value when one of the dependencies has changed`. and if you use the `getUserName` function as a dependency, the functino itself will never change. this won't work.

Also `username` is a string and most likely a constant stored in provider. when `getUsername` gets called every time, it will return the same value.

Reply:
I can see the problem when username is populated after the provider class is initiated,e.g. async all to get username information.
Otherwise, if both getUserName and the actual value of the username were to be initiated at the same time and never change, e.g. in case of a constant, this would still work.
We need to make sure we support the async scenario as well, therefore, you are right on not using useMemo

Reply:
right, username is only assigned when the user successfully logged in